pyacq/core/stream/stream.py

```diff
@@ -0,0 +1,294 @@
+        else:
+            self.url = '{protocol}://{interface}:{port}'.format(**self.params)
+            
+        # allow some keys in self.spec to override self.params
+        readonly_params = ['protocol', 'transfermode', 'shape', 'dtype']
+        for k,v in self.spec.items():
+            if k in readonly_params and v != self.params[k]:
+                raise ValueError("InputStream parameter %s=%s does not match connected output %s=%s." %
+                                 (k, v, k, self.params[k]))
+            else:
+                self.params[k] = v
+
+        context = zmq.Context.instance()
+        self.socket = context.socket(zmq.SUB)
+        self.socket.setsockopt(zmq.SUBSCRIBE, b'')
+        #~ self.socket.setsockopt(zmq.DELAY_ATTACH_ON_CONNECT,1)
+        self.socket.connect(self.url)
+        
+        transfermode = self.params['transfermode']
+        if transfermode not in all_transfermodes:
+            raise ValueError("Unsupported transfer mode '%s'" % transfermode)
+        receiver_class = all_transfermodes[transfermode][1]
+        self.receiver = receiver_class(self.socket, self.params)
+        
+        self.connected = True
+        if self.node and self.node():
+            self.node().after_input_connect(self.name)        
+    
+    def poll(self, timeout=None):
+        """Poll the socket of input stream.
+        """
+        return self.socket.poll(timeout=timeout)
+    
+    def recv(self, **kargs):
+        """
+        Receive a chunk of data.
+        
+        Returns:
+        ----
+        index: int
+            The absolute sample index. If the chunk contains multiple samples,
+            then this is the index of the last sample.
+        data: np.ndarray or bytes
+            The received chunk of data.
+            If the stream uses `transfermode='sharedarray'`, then the data is 
+            returned as None and you must use `InputStream.get_array_slice(index, length)`
+            to read from the shared array or InputStream.recv(with_data=True) to get the last
+            chunk.
+
+        """
+        index, data = self.receiver.recv(**kargs)
+        if self._own_buffer and data is not None and self.buffer is not None:
+            self.buffer.new_chunk(data, index=index)
+        return index, data
+
+    def close(self):
+        """Close the Input.
+        
+        This closes the socket.
+        
+        """
+        self.receiver.close()
+        self.socket.close()
+        del self.socket
+    
+    def __getitem__(self, *args):
+        """Return a data slice from the RingBuffer attached to this InputStream.
+        
+        If no RingBuffer is attached, raise an exception. See ``set_buffer()``.
+        """
+        if self.buffer is None:
+            raise TypeError("No ring buffer configured for this InputStream.")
+        return self.buffer.__getitem__(*args)
+
+    def set_buffer(self, size=None, double=True, axisorder=None):
```

Comment:
HMMMM. Maybe not; I think this actually makes the oscilloscope code more complex.

In general, the oscilloscope should not know or care what kind of stream it is connected to; it always needs to ensure that a ring buffer of sufficient size is availaable. If it is connected to plaindata, then a new ring buffer needs to be created. If it is connected to sharedmem but the shared buffer is too small, then again a new buffer needs to be created. 

So rather than have the oscilloscope (and every other input node) check to see whether a large-enough buffer is available, we just have it declare, "I need a ring buffer with these properties" and let the inputstream decide whether to create a new one.